### PR TITLE
Rust core formula store

### PR DESCRIPTION
**Pull request by davidkircos**

### CODE REVIEW

File: quadratic-core/src/grid/controller.rs [hunk 1]
```diff
@@ -326,21 +328,25 @@ impl GridController {
...
+                Command::SetCellCode(p1, old_value)
             }
         };
 
         Ok(reverse_command)
     }
 
+    pub fn get_grid(&self) -> &Grid {
+        &self.grid
+    }
```
Comment (davidkircos) [inline quadratic-core/src/grid/controller.rs:348]: Is there a better way to get a read only ref to grid from GridController?

File: quadratic-core/src/tests/gc_deps.rs [hunk 2]
```diff
@@ -111,9 +114,23 @@ fn test_gc_simulate_user_actions() {
 
     // User sets B1 to "A1 + A2"
     grid.transact(|t| {
-        // TODO: Command to save formula on B1 of "=A1+A2" in R1C1 form "=RC[-1]+R[1]C[-1]"
-        // TODO: Compute result
-        t.exec(Command::SetCell(Pos { x: 1, y: 0 }, Cell::Int(30)))?;
+        // eval formula
+        let form = Formula::new_sum(&[Pos { x: 0, y: 0 }, Pos { x: 0, y: 1 }]);
+        let r_cell = form.eval(t.grid().get_grid(), Pos { x: 1, y: 0 }).unwrap();
```
Comment (davidkircos) [inline quadratic-core/src/tests/gc_deps.rs:119]: This is sort of an awkward call `t.grid().get_grid()`

Maybe the first function should be renamed `t.grid_controller().grid()` ?

**Review by HactarCE — CHANGES_REQUESTED** [review]

File: quadratic-core/src/codestore/store.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct CodeStore {
+    pub code: HashMap<Pos, CodeCell>,
+}
+
+impl fmt::Display for CodeStore {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "CodeStore {{ code: {:?} }}", self.code)
```
[the remark this answers is not in the document]
Comment (davidkircos) [reply quadratic-core/src/codestore/store.rs:27]: Removed Display impl

File: quadratic-core/src/codestore/store.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        Self::default()
+    }
+
+    pub fn set_cell_code(
+        &mut self,
+        pos: Pos,
+        code_cell_or_none: Option<CodeCell>,
+    ) -> Option<CodeCell> {
+        // sets code at pos and returns the old code or None if there was no code at pos
+        if code_cell_or_none.is_some() {
```
[the remark this answers is not in the document]
Comment (davidkircos) [reply quadratic-core/src/codestore/store.rs:43]: can you show me what you mean by `if let Some(...) = ...`

[the remark this answers is not in the document]
Comment (HactarCE) [reply quadratic-core/src/codestore/store.rs]:
```rs
if let Some(code_cell) = code_cell_or_none {
    self.code.insert(pos, code_cell)
} else {
    self.code.remove(&pos)
}
```

**Pull request merged**